File: A2Z_striver/1 Basics/1.2 Patterns/pattern_8_m/pattern_8.py
import sys

class Solution:
    """
    Problem: Print an Inverted Star Pyramid.
    Example Input: n = 3
    Example Output:
    *****
     ***
      *
    
    Constraints: 1 <= n <= 100
    Goal: O(N^2) Time.
    """

    # // ~ Approach 1: Nested Loops (Algorithmic Standard)
    # // Analysis: Time O(N^2) | Space O(1)
    def pattern8_nested(self, n):
        for i in range(n):
            # // - Step 1: Print Leading Spaces
            # //   Logic: 0 spaces, then 1, then 2... (equal to row index)
            for j in range(i):
                print(" ", end="")

            # // - Step 2: Print Stars
            # //   Logic: Decreasing odd numbers.
            # //   i=0 -> 2*n - 1
            # //   i=1 -> 2*(n-1) - 1
            for j in range(2 * (n - i) - 1):
                print("*", end="")
            
            print()

    # // ~ Approach 2: Pythonic One-Liner
    # // Analysis: Time O(N^2) | Space O(1)
    def pattern8_pythonic(self, n):
        for i in range(n):
            # // . Direct string math
            spaces = " " * i
            stars = "*" * (2 * (n - i) - 1)
            print(spaces + stars)

# Driver Code
if __name__ == "__main__":
    solver = Solution()
    N = 5
    
    print(f"--- Approach 1: Nested Loops (N={N}) ---")
    solver.pattern8_nested(N)

File: A2Z_striver/1 Basics/1.2 Patterns/pattern_8_m/pattern_8.cpp
#include <iostream>

using namespace std;

// Header Comment Block
// Problem: Print an Inverted Star Pyramid.
// Example Input: n = 3
// Example Output:
// *****
//  ***
//   *
// Goal: O(N^2) Time | O(1) Space

class Solution
{
public:
    // ~ Approach: Dependent Nested Loops (Inverted Logic)
    // Analysis: Time O(N^2) | Space O(1)
    void pattern8(int n)
    {
        for (int i = 0; i < n; i++)
        {

            // - Step 1: Print Spaces
            // - Logic: Indentation increases with row index (0, 1, 2...)
            for (int j = 0; j < i; j++)
            {
                cout << " ";
            }

            // - Step 2: Print Stars
            // - Logic: Odd numbers decreasing (2n-1, 2n-3...)
            for (int j = 0; j < 2 * (n - i) - 1; j++)
            {
                cout << "*";
            }

            cout << "\n";
        }
    }

    void pattern8StrictTwoLoops(int n) {
        // Outer Loop: Iterates through rows (0 to n-1)
        for (int i = 0; i < n; i++) {
            
            // 1. Calculate limits for this row
            // Spaces logic: Row 0 -> 0 spaces, Row 1 -> 1 space... (spaces = i)
            int spaces = i;
            
            // Stars logic: 2*n - (2*i + 1) OR 2*(n-i) - 1
            int stars = 2 * (n - i) - 1;
            
            // The inner loop must run for the sum of both
            int totalCols = spaces + stars;

            // Inner Loop: STRICTLY ONE LOOP for columns
            for (int j = 0; j < totalCols; j++) {
                if (j < spaces) {
                    cout << " "; // Print Space
                } else {
                    cout << "*"; // Print Star
                }
            }
            
            cout << "\n";
        }
    }

};

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    Solution sol;
    int n = 5;

    cout << "Test Case (n=" << n << "):\n";
    sol.pattern8(n);

    return 0;
}